pkg/generator/gen.go

```diff
@@ -0,0 +1,340 @@
+	if err := validateFlags(f); err != nil {
+		return err
+	}
+
+	if err := g.setDefaults(f); err != nil {
+		return err
+	}
+	return g.generate(ctx)
+}
+
+// validateFlags checks if the inputs provided through flags are valid and
+// if so, sets defaults.
+func validateFlags(f flag.Flags) error {
+	if f.InputDir == "" {
+		return errors.New("input path to API definition is required.")
+	}
+
+	if f.ClientsetAPIPath == "" {
+		return errors.New("specifying client API path is required currently.")
+	}
+
+	// TODO: Do we default this from name of the type?
+	if f.InterfaceName == "" {
+		return errors.New("specifying interface name is required currently.")
+	}
+
+	if len(*f.GroupVersions) == 0 {
+		return errors.New("list of group versions for which the clients are to be generated is required.")
+	}
+
+	return nil
+}
+
+// setDefaults sets the default values for the generator. It also creates
+// a list of group versions provided as an input.
+func (g *Generator) setDefaults(f flag.Flags) error {
+	if f.InputDir != "" {
+		g.inputDir = f.InputDir
+	}
+	if f.OutputDir != "" {
+		g.outputDir = f.OutputDir
+	}
+	if f.ClientsetAPIPath != "" {
+		g.clientSetAPIPath = f.ClientsetAPIPath
+	}
+	if f.InterfaceName != "" {
+		g.interfaceName = f.InterfaceName
+	}
+	return g.getGV(f)
+}
+
+// getGV parses the Group Versions provided in the input through flags
+// and creates a list of []types.GroupVersions.
+// Note: Though each type.GroupVersions is allowed to have multiple versions,
+// we are restricting it to one currently.
+// Hence, for ex: only "rbac:v1" is accepted, instead of "rbac:v1,v2,v3".
+func (g *Generator) getGV(f flag.Flags) error {
+	// Its already validated that list of group versions cannot be empty.
+	gvs := *f.GroupVersions
+	for _, gv := range gvs {
+		// arr[0] -> group, arr[1] -> version
+		arr := strings.Split(gv, ":")
+		if len(arr) != 2 {
+			return fmt.Errorf("input to --group-version must be in <group>:<version> format, ex: rbac:v1. Got %q", gv)
+		}
+
+		// input path is converted to <inputDir>/pkg/apis/<group>/<version>.
+		// example for input directory of "k8s.io/client-go/kubernetes", it would
+		// be converted to "k8s.io/client-go/kubernetes/pkg/apis/rbac/v1".
+		input := filepath.Join(f.InputDir, "pkg", "apis", arr[0], arr[1])
+		groups := []types.GroupVersions{}
+		builder := args.NewGroupVersionsBuilder(&groups)
+		_ = args.NewGVPackagesValue(builder, []string{input})
+
+		g.groupVersions = append(g.groupVersions, groups...)
+	}
+	return nil
+}
+
+// generate first genrates the wrapper for all the interfaces provided in the input.
+// Then for each type defined in the input, it recursively wraps the subsequent
+// interfaces to be kcp-aware.
+func (g *Generator) generate(ctx *genall.GenerationContext) error {
+	err := g.writeWrappedClientSet()
+	if err != nil {
+		return err
+	}
+	return g.generateSubInterfaces(ctx)
+}
+
+func (g *Generator) writeWrappedClientSet() error {
+	out := new(bytes.Buffer)
```

Comment:
Global comment: the godoc indicates `The zero value for Buffer is an empty buffer ready to use.`, so I think you could do `var out bytes.Buffer` to avoid making it a pointer. LMK if that works or not.